Bugfix/issue 320 fix tsan

yokofly opened the pull request:
_edit v1: update more pr info_ 

Please write user-readable short description of the changes:
fix #320 
stacktrace here:
1. when quit proton-server muduo timeservice data race [stack trace here](https://github.com/timeplus-io/proton/issues/320#issuecomment-1821966387)
2. run smoke test: docheckpoint data race [stack trace here](https://github.com/timeplus-io/proton/issues/320#issuecomment-1821969630)



useful link:
https://en.cppreference.com/w/cpp/atomic/atomic/wait
https://www.youtube.com/watch?v=gTpubZ8N0no

yokofly committed:
fix thread sanitizer with extra link library.

yokofly committed:
add extra mutex to fix data race in timerservice.

maybe there can be other simple methods with atomic var usage(cpp20).

yokofly committed:
use scoped_lock instead lock_guard in TimerService

yokofly committed:
replace atomic<ckpt_ctx*> with shared_ptr and mutex lock

yokofly committed:
Merge branch 'develop' into bugfix/issue-320-fix-tsan

yl-lisen commented on src/Common/TimerService.h:
```diff
@@ -26,6 +28,7 @@ class TimerService final
     /// Safe to call from other threads.
     muduo::net::TimerId runAfter(double delay, muduo::net::TimerCallback cb)
     {
+        std::scoped_lock lock(event_loop_mutex);
         assert(event_loop);
         return event_loop->runAfter(delay, std::move(cb));
```
After TimerService shutdown, the event_loop is nullptr, but it still may be called in `CheckpointCoordinator::removeExpiredCheckpoints` ?

yl-lisen replied:
We can make the class `TimerService` is thread safe, or we can also ensure it where used

yokofly replied:
that's the reason I added a lock guard here, and checked the pointer exists in the shutdown function.
Do you mean we need to check the pointer exists before calling the function in the `runAfter` ? 
```
void TimerService::shutdown()
{
    {
        std::scoped_lock lock(event_loop_mutex);
        if (event_loop)
        {
            event_loop->quit();

            /// Decrement ref count to make sure the last ref count
            /// of event loop is in the creation thread
            event_loop = nullptr;
        }
    }

    eloop_init_guard.store(nullptr);
    eloop_init_guard.notify_all();

    looper.wait();
}
```

yl-lisen replied:
yes, as above mentioned that when quit proton-server muduo timeservice data race https://github.com/timeplus-io/proton/issues/320#issuecomment-1821966387 ,
One thread is calling  `TimerService::shutdown()` in `~CheckpointCoordinator()`
Another thread is calling `TimerService::runAfter()` (belongs to EventLoop thread)

yokofly committed:
fix review comment: throw if eventloop not exist

yokofly committed:
Merge branch 'develop' of github.com:timeplus-io/proton into bugfix/issue-320-fix-tsan

yl-lisen commented on src/Storages/Streaming/StreamingStoreSourceBase.cpp:
```diff
@@ -161,13 +168,14 @@ Chunk StreamingStoreSourceBase::generate()
 {
     /// We assume the previous ckpt is already done
     /// Use std::atomic<std::shared_ptr<CheckpointContext>>
-    assert(!ckpt_ctx.load(std::memory_order_relaxed));
-    ckpt_ctx = new CheckpointContext(*ckpt_ctx_);
+    std::scoped_lock lock(ckpt_mutex);
+    assert(!ckpt_ctx);
+    ckpt_ctx = std::make_shared<CheckpointContext>(*ckpt_ctx_);
```
I want to know why the previous one was wrong?

yokofly replied:
https://github.com/timeplus-io/proton/issues/320#issuecomment-1821969630
one thread alloc/write by `new`
one thread is access/read with `doCheckpoint`

yl-lisen replied:
It's an atomic variable, so why does this problem arise?

yokofly replied:
I try many times in my local with my assumption, but do not get a thread sanitizer report(hahaha).

basically, I believe the problem is `ckpt_ctx = new CheckpointContext(*ckpt_ctx_);`
this phase has done 2 things:
the first step is new a object, then the second step will be atomically assigned.

the key is: thread 2(t2) may read old_obj, in the same time ckpt_ctx has been assigning with new_obj in thread 1(t1)
the pointer is safe, but the pointer reference object is not guaranteed. 
with a mutex lock guard, the  t1's new and assign operations are seen as a single, atomic action from t2's perspective.

```
// from gpt
[Thread 1 - Writer]                              [Thread 2 - Reader]
   |                                                   |
   |--- (1) new_obj = new CheckpointContext(...) -----|  // Creates a new object (deep copy)
   |                                                   |
   |--- (2) ckpt_ctx = new_obj -----------------------|  // Atomically assigns the new object's
   |        (Atomic assignment)                        |  // address to ckpt_ctx.
   |                                                   |
   |===================== DATA RACE ===================|
   |                                                   |
   |                                                   |--- (3) current = ckpt_ctx.exchange(nullptr)
   |                                                   |     (Atomic exchange)
   |                                                   |
   |--- (4) Return from function ----------------------|  // Thread 1 completes its operation.
   |                                                   |     // Thread 2 may now be working with
   |                                                   |     // the object previously pointed to by
   |                                                   |     // ckpt_ctx.

```

yokofly committed:
fix compile-add extern errcode

yokofly commented:
may follow up https://github.com/timeplus-io/proton/pull/111

chenziliang commented on src/Common/TimerService.cpp:
```diff
@@ -12,12 +12,19 @@ void TimerService::startup()
+            /// Decrement ref count to make sure the last ref count
+            /// of event loop is in the creation thread
+            event_loop = nullptr;
+        }
```
if event_loop is nullptr, we shall return directly ?

chenziliang commented on src/Common/TimerService.h:
```diff
@@ -18,31 +24,47 @@ class TimerService final
     muduo::net::TimerId runAt(muduo::Timestamp time, muduo::net::TimerCallback cb)
     {
-        assert(event_loop);
+        std::scoped_lock lock(event_loop_mutex);
```
I think we shall avoid using lock here since runAt is thread safe. Even event_loop itself is not thread safe, we require users to align `shutdown` and `runAt` which means shutdown is only invoked when system is going down and after that we shall never call `runAt`. If we do, there is logic error.

chenziliang replied:
This is why i ask if we can just mask TSAN for TimerService ? which means don't apply TSA to TimerService since we know what we are doing and for now, it seems not quite worth fixing it since we will decommission it.

chenziliang commented on src/Storages/Streaming/StreamingStoreSourceBase.h:
```diff
@@ -50,7 +50,8 @@ class StreamingStoreSourceBase : public ISource
-    /// FIXME, switch to llvm-15
-    std::atomic<CheckpointContext *> ckpt_ctx;
+    /// FIXME, switch to llvm-15 atomic shared_ptr
+    CheckpointContextPtr ckpt_ctx TSA_PT_GUARDED_BY(ckpt_mutex);
```
1. In general (general guideline), `mutex` shall be defined before the data members it protects since when dtor, data members will be dtor first and then mutex and this is safer in different scenarios. 
2. There is race condition, this is exactly why we have a `FIXME`. Can you try if `std::atomic<CheckpointContextPtr>` works ? And for now, we tolerate this race because they are usually scheduled at least seconds if not mins away. So in practice, it may not quite matter. Having said this, can you try `std::atomic<CheckpointContextPtr>`

chenziliang replied:
We are trying to avoid a mutex here since it is in hot path.

yokofly replied:
_edit v1: update the compiler link_
I know, but even in clang-17 the compiler still does not support it.

see below the online compiler:
https://godbolt.org/z/aGaEPbEc7
or check this compiler support in cppref:
https://en.cppreference.com/w/cpp/compiler_support#:~:text=Atomic%20std%3A%3Ashared_ptr,19.27*

yokofly committed:
Merge remote-tracking branch 'origin/develop' into bugfix/issue-320-fix-tsan

yokofly committed:
update review comment:
1. get rid of lock usage in timerservice
2. add suppression in diskutil check

yokofly committed:
Merge remote-tracking branch 'origin/develop' into bugfix/issue-320-fix-tsan

yokofly committed:
Merge remote-tracking branch 'origin/develop' into bugfix/issue-320-fix-tsan

chenziliang commented on src/Storages/Streaming/StreamingStoreSourceBase.h:
```diff
@@ -50,7 +50,8 @@ class StreamingStoreSourceBase : public ISource
-    std::atomic<CheckpointContext *> ckpt_ctx;
+    /// FIXME, switch to llvm-15 atomic shared_ptr
+    CheckpointContextPtr ckpt_ctx TSA_PT_GUARDED_BY(ckpt_mutex);
+    std::mutex ckpt_mutex;
```
We may need apply 
1) fast mutex 
2) double checked locking to avoid lock a mutex always.

```
FastMutext ckpt_mutex;
std::atomic<bool> has_ckpt_request = false;
```

```
/// Check
{
    CheckpointContextPtr ckpt;
    if (has_ckpt_request)
    {
        std::scope_lock lock(ckpt_mutex);
        if (ckpt_ctx)
        {
            ckpt.swap(ckpt_ctx); <-- use swap probably faster ?
            has_ckpt_request = false;
        }
    }
}


/// Set
{
    std::scope_lock lock(ckpt_mutex);
    /// ckpt coordinator guarantees that ckpt is not overlapped.
    assert(!has_ckpt_request);
    assert(!ckpt_ctx);
    ckpt_ctx = std::make_shared<CheckpointContext>(*ckpt_ctx_);
    has_ckpt_request = true;
}
```

yokofly replied:
this is just a demo about mutex and atomicbool check.not ready merge.

micro benchmark I will
do in later . 

I intend to use atomic<ckpt*> and std mutex.

yokofly committed:
update with lockfree+checks
first check atomic var, then get lock mutex.

yokofly committed:
use atomic_bool inmemory in nativelog.

yokofly committed:
Merge remote-tracking branch 'origin/develop' into bugfix/issue-320-fix-tsan

yokofly committed:
add atomic lock double check perf
```
$ ./src/Storages/Streaming/benchmarks/atomic_lock_double_check_perf
2023-12-04T21:57:55-08:00
Running ./src/Storages/Streaming/benchmarks/atomic_lock_double_check_perf
Run on (20 X 4900 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x10)
  L1 Instruction 32 KiB (x10)
  L2 Unified 1280 KiB (x10)
  L3 Unified 24576 KiB (x1)
Load Average: 4.06, 4.53, 4.08
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
------------------------------------------------------------------------------------------------
Benchmark                                                      Time             CPU   Iterations
------------------------------------------------------------------------------------------------
BM_ScopedLockOnly/min_time:3.000                            17.3 ns         17.3 ns    238669686
BM_AtomicWithScopedLockOnStdMutex/min_time:3.000           0.270 ns        0.269 ns   1000000000
BM_AtomicWithScopedLockOnPocoFastMutex/min_time:3.000      0.279 ns        0.275 ns   1000000000
$ ./src/Storages/Streaming/benchmarks/atomic_lock_double_check_perf
2023-12-04T21:59:34-08:00
Running ./src/Storages/Streaming/benchmarks/atomic_lock_double_check_perf
Run on (20 X 4900 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x10)
  L1 Instruction 32 KiB (x10)
  L2 Unified 1280 KiB (x10)
  L3 Unified 24576 KiB (x1)
Load Average: 4.40, 4.40, 4.07
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
------------------------------------------------------------------------------------------------
Benchmark                                                      Time             CPU   Iterations
------------------------------------------------------------------------------------------------
BM_ScopedLockOnly/min_time:3.000                            21.0 ns         20.9 ns    175392202
BM_AtomicWithScopedLockOnStdMutex/min_time:3.000           0.265 ns        0.265 ns   1000000000
BM_AtomicWithScopedLockOnPocoFastMutex/min_time:3.000      0.264 ns        0.264 ns   1000000000
$
```
there is no much diff in poco::mutex and std::mutex on this case. i intend to use std::mutex to make code more 'standard'

yokofly committed:
apply std::mutex plus atomic_bool double check style
it hard to lock-free with atomic<shared_ptr> even with potential future libcpp impl

yokofly committed:
revert header style change in timerservice. make code review a bit more easy

yokofly committed:
Merge remote-tracking branch 'origin/develop' into bugfix/issue-320-fix-tsan

yokofly committed:
Merge branch 'develop' into bugfix/issue-320-fix-tsan

yokofly committed:
Merge branch 'develop' into bugfix/issue-320-fix-tsan

chenziliang commented on src/Storages/Streaming/StreamingStoreSourceBase.cpp:
```diff
@@ -161,13 +171,20 @@ Chunk StreamingStoreSourceBase::generate()
-    assert(!ckpt_ctx.load(std::memory_order_relaxed));
-    ckpt_ctx = new CheckpointContext(*ckpt_ctx_);
+    assert(!has_ckpt_request);
+    CheckpointContextPtr new_ckpt = std::make_shared<CheckpointContext>(*ckpt_ctx_);
```
Do we need make a new_ckpt ?

yokofly replied:
i got many many error about the ckpt file path duplicate(i assume this is due to the shallow copy)
work with deep copy everything well.

chenziliang replied:
Could you elaborate the error ? When it happened and what's the exact error message ?

yokofly replied:
it works with shallow copy today.. but I remembered yesterday if I do not take the deep copy, the smoke test will break.
and I find the original `atomic<ckpt_ctx*>` also does deep copy, thus I take this trial.

chenziliang commented on src/Storages/Streaming/StreamingStoreSourceBase.cpp:
```diff
@@ -138,8 +138,18 @@ Chunk StreamingStoreSourceBase::generate()
+            std::scoped_lock lock(ckpt_mutex);
+            current.swap(ckpt_ctx);
+        }
+        has_ckpt_request = false;
```
We will need set it under ckpt_mutex held.

yokofly replied:
why? Does Atomic Value also need a mutex guard?

yokofly replied:
discussed internally:
```
This is a race case.
Thr1: check if having ckpt req -> true -> hold mutex -> swap ckpt -> unlock mutex => at this very time schedule ckpt set thread
Thr2: set checkpoint ctx -> hold mutex -> swap ctx -> unlock mutex -> set has_req bool to true => context switch to Thr1
Thr1: -> set the has_req bool to false. Now we missed handling one request context req
```

chenziliang commented on src/Storages/Streaming/StreamingStoreSourceBase.cpp:
```diff
@@ -138,8 +138,18 @@ Chunk StreamingStoreSourceBase::generate()
 
-    if (auto * current = ckpt_ctx.exchange(nullptr, std::memory_order_relaxed); current)
-        return doCheckpoint(CheckpointContextPtr{current});
+    CheckpointContextPtr current;
```
`current` -> `current_ckpt_ctx`

chenziliang commented on src/Storages/Streaming/StreamingStoreSourceBase.cpp:
```diff
@@ -138,8 +138,18 @@ Chunk StreamingStoreSourceBase::generate()
+    }
+
+    if (current)
+        return doCheckpoint(current);
```
`std::move(current)`

chenziliang commented on src/Storages/Streaming/StreamingStoreSourceBase.cpp:
```diff
@@ -161,13 +171,20 @@ Chunk StreamingStoreSourceBase::generate()
+        assert(!ckpt_ctx);
+        ckpt_ctx.swap(new_ckpt);
+    }
+    has_ckpt_request = true;
```
We will need set it with ckpt_mutex held.

chenziliang commented on src/Storages/Streaming/benchmarks/atomic_lock_double_check_perf.cpp:
```diff
@@ -0,0 +1,145 @@
+    [[maybe_unused]] int valid_num;
+};
+
+static void BM_ScopedLockOnly(benchmark::State & state)
```
Consider anonymous namespace instead of static

chenziliang replied:
like 

```
namespace
{
..
}

BENCHMARK(BM_ScopedLockOnly)->MinTime(3);
BENCHMARK(BM_AtomicWithScopedLockOnPocoFastMutex)->MinTime(3);
```

yokofly replied:
i see. tks

chenziliang commented on src/NativeLog/Log/Log.h:
```diff
@@ -304,7 +304,7 @@ class Log final : private boost::noncopyable
     std::atomic_flag closed;
 
-    bool inmemory = false;
+    std::atomic<bool> inmemory = false;
```
Why this need to be atomic since it is set once during ctor and read-only afterwards ?

yokofly replied:
the boolean var also report by thread sanitizer,  basically the usage(read) does not been guarantee the thread safety.

chenziliang replied:
Interesting. Can we mask this variable to avoid the TSA ?

yokofly replied:
https://github.com/timeplus-io/proton/issues/320#issuecomment-1842414990

chenziliang commented on src/Storages/ExternalStream/Kafka/KafkaSource.h:
```diff
@@ -87,8 +87,10 @@ class KafkaSource final : public ISource
     /// For checkpoint
-    /// FIXME, switch to llvm-15
-    std::atomic<CheckpointContext *> ckpt_ctx;
+    /// FIXME, switch to llvm-15 atomic shared_ptr
```
We can consider refactor it out as a util struct like below and reuse everywhere ?

```
/// \CheckpointRequest applies double check locking to minimize using mutex 
/// directly for better performance
class CheckpointRequest
{
public:
    /// \poll moves out the existing checkpoint context request atomically if there is any.
    CheckpointContextPtr poll() noexcept 
    { 
        CheckpointContextPtr new_request;
        if (has_ckpt_request)
        {
            std::scoped_lock lock(ckpt_mutex);
            new_request.swap(ckpt_ctx);
            has_ckpt_request = false;
        }
        return new_request;
    }

     void setCheckpointRequestCtx(CheckpointContextPtr ckpt_ctx_) noexcept
     {
         assert(!has_ckpt_request);
         std::scoped_lock lock(ckpt_mutex);
         assert(!ckpt_ctx);
         ckpt_ctx.swap(ckpt_ctx_);
         has_ckpt_request = true;
     }

     bool hasRequest() const noexcept { return has_ckpt_request; }

private:
    std::atomic<bool> has_ckpt_request = false;
    std::mutex ckpt_mutex;
    CheckpointContextPtr ckpt_ctx TSA_PT_GUARDED_BY(ckpt_mutex);
};
```

chenziliang replied:
Place this util class in the same directory as where `CheckpointContext` is defined

yokofly committed:
fix bench by anonymous namespace instead static fn

yokofly committed:
fix review comment:
1. nlog use boolean instead atomic<bool>
2. extra ckpt_request in a common util

yokofly committed:
fix build by use runtime suppression instead blacklist

chenziliang commented on src/Checkpoint/CheckpointRequest.h:
```diff
@@ -0,0 +1,54 @@
+    {
+        assert(!has_ckpt_request);
+        /// Perform a deep copy to ensure that the checkpoint file is stored in a distinct path.
+        CheckpointContextPtr new_ckpt = std::make_shared<CheckpointContext>(*ckpt_ctx_);
```
I don't think we will need a deep copy ?

yokofly replied:
it works with shallow copy today.. but I remembered yesterday if I do not take the deep copy, the smoke test will break.
and I find the original atomic<ckpt_ctx*> also does deep copy, thus I take this trial.

yokofly committed:
Truly, yesterday it cannot work, but it works well today.
maybe I mixed the build folder or proton data folder<
* get rid of deep copy for ckpt_ctx
* move epoch calculate into necessary condition.(very small optimization)
